crates/core/tedge_api/src/workflow/mod.rs

```diff
@@ -270,24 +270,33 @@ impl OperationAction {
     pub fn inject_state(&self, state: &GenericCommandState) -> Self {
         match self {
             OperationAction::Script(script, handlers) => OperationAction::Script(
-                ShellScript {
-                    command: state.inject_parameter(&script.command),
-                    args: state.inject_parameters(&script.args),
-                },
+                Self::inject_values_into_script(state, script),
                 handlers.clone(),
             ),
-            OperationAction::Operation(command, Some(script), input, handlers) => {
+            OperationAction::BgScript(script, handlers) => OperationAction::BgScript(
+                Self::inject_values_into_script(state, script),
+                handlers.clone(),
+            ),
+            OperationAction::Operation(operation_expr, optional_script, input, handlers) => {
+                let operation = state.inject_values_into_template(operation_expr);
+                let optional_script = optional_script
+                    .as_ref()
+                    .map(|script| Self::inject_values_into_script(state, script));
                 OperationAction::Operation(
-                    command.clone(),
-                    Some(ShellScript {
-                        command: state.inject_parameter(&script.command),
-                        args: state.inject_parameters(&script.args),
-                    }),
+                    operation,
+                    optional_script,
                     input.clone(),
                     handlers.clone(),
                 )
             }
             _ => self.clone(),
         }
     }
+
+    fn inject_values_into_script(state: &GenericCommandState, script: &ShellScript) -> ShellScript {
+        ShellScript {
+            command: state.inject_values_into_template(&script.command),
```

Comment:
Jus wondering if we should add some additional restrictions to the variable substitution in script and operation names, limiting the substitution to primitive types only (Strings, numbers etc) and explicitly prevent complex values like arrays and objects as that's most likely an error when used in a script/operation name? It would be better fail during the validation phase itself rather than during execution.

Reply:
Yeah, why not.

Reply:
Though this shouldn't prevent passing complex objects as arguments to the script right?

For example:

```toml
[executing]
script = "/usr/bin/do_something ${.payload}"
```

Reply:
Sure. I think @albinsuresh is speaking only about operation names.